### PR TITLE
nsswitch: Add try_authok option to pam_winbind

### PR DESCRIPTION
**Pull request by sathieu**
Same as the use_authtok option, except that if the new password is not
valid, PAM will prompt for a password.

Bug-Debian: https://bugs.debian.org/858923
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/samba/+bug/570944

Signed-off-by: Mathieu Parent <math.parent@gmail.com>

### CODE REVIEW

**Comment by cryptomilk** [comment]
Could you please add a test for this?

See for example:
python/samba/tests/pam_winbind_warn_pwd_expire.py

and

https://cwrap.org/pam_wrapper.html

**Comment by sathieu** [comment]
Yes. But how to add a "previous module" like pam_unix?

**Comment by cryptomilk** [comment]
I think you just need to set a PAM env variable right? pam_wrapper provides a module pam_set_items which also has a manpage: 'man pam_set_items'. You can use it so set the password pam env.

**Comment by sathieu** [comment]
Unfortunately pam_set_items is not in `third_party/pam_wrapper/`. I don't know enough of waf to add it.

**Comment by cryptomilk** [comment]
This looks promising!

**Comment by sathieu** [comment]
I don't understand why the `echo_off_list` (`[password, newpassword, newpassword]`) is not passed to pam_winbind. Any clues?

Also, as those tests change the password, how should we handle password policy (i.e min password age)?

**Comment by sathieu** [comment]
Here is the failing log (`could not recover authentication token`, in `pam_winbind.c` line 2229):
```
[1(0)/2 at 0s] samba.tests.pam_winbind with options 'try_authtok'(local)(ad_member)
PWRAP_DEBUG(29944) - pwrap_init: Initialize pam_wrapper
PWRAP_DEBUG(29944) - pwrap_init: Copy /usr/lib/x86_64-linux-gnu/libpam.so.0.83.1 to /tmp/pam.f/lib/libpam.so.0
PWRAP_DEBUG(29944) - copy_confdir: Copy config files from /home/mathieu/src/git.samba.org/samba/st/tmp/pam_services to /tmp/pam.f
PWRAP_DEBUG(29944) - pwrap_init: Successfully initialized pam_wrapper
PWRAP_DEBUG(29945) - pwrap_init: Initialize pam_wrapper
PWRAP_DEBUG(29945) - pwrap_init: Copy /usr/lib/x86_64-linux-gnu/libpam.so.0.83.1 to /tmp/pam.g/lib/libpam.so.0
PWRAP_DEBUG(29945) - copy_confdir: Copy config files from /home/mathieu/src/git.samba.org/samba/st/tmp/pam_services to /tmp/pam.g
PWRAP_DEBUG(29945) - pwrap_init: Successfully initialized pam_wrapper
PWRAP_DEBUG(29945) - pwrap_load_lib_handle: Opened /tmp/pam.g/lib/libpam.so.0

PWRAP_DEBUG(29945) - SYSLOG(7): [pamh: 0x55a119be8200] ENTER: pam_sm_chauthtok (flags: 0x4000)
PWRAP_DEBUG(29945) - SYSLOG(7): [pamh: 0x55a119be8200] STATE: ITEM(PAM_SERVICE) = "samba" (0x55a119987fe0)
PWRAP_DEBUG(29945) - SYSLOG(7): [pamh: 0x55a119be8200] STATE: ITEM(PAM_USER) = "LOCALADMEMBER/root" (0x55a1199880b0)
PWRAP_DEBUG(29945) - SYSLOG(7): [pamh: 0x55a119be8200] STATE: ITEM(PAM_CONV) = 0x55a119988180
PWRAP_DEBUG(29945) - SYSLOG(7): username [LOCALADMEMBER/root] obtained
PWRAP_DEBUG(29945) - SYSLOG(7): getting password (0x00009021)
PWRAP_DEBUG(29945) - SYSLOG(5): could not recover authentication token
PWRAP_DEBUG(29945) - SYSLOG(7): unable to obtain a password
PWRAP_DEBUG(29945) - SYSLOG(5): password - (old) token not obtained
PWRAP_DEBUG(29945) - SYSLOG(7): [pamh: 0x55a119be8200] LEAVE: pam_sm_chauthtok returning 21 (PAM_AUTHTOK_RECOVER_ERR)
PWRAP_DEBUG(29945) - SYSLOG(7): [pamh: 0x55a119be8200] STATE: ITEM(PAM_SERVICE) = "samba" (0x55a119987fe0)
PWRAP_DEBUG(29945) - SYSLOG(7): [pamh: 0x55a119be8200] STATE: ITEM(PAM_USER) = "LOCALADMEMBER/root" (0x55a1199880b0)
PWRAP_DEBUG(29945) - SYSLOG(7): [pamh: 0x55a119be8200] STATE: ITEM(PAM_CONV) = 0x55a119988180
UNEXPECTED(error): samba.tests.pam_winbind with options 'try_authtok'(local).samba.tests.pam_winbind_options.PamOptionsTests.test_chauthtok(ad_member)
REASON: Exception: Exception: Traceback (most recent call last):
  File "/home/mathieu/src/git.samba.org/samba/bin/python/samba/tests/pam_winbind_options.py", line 34, in test_chauthtok
    res = pypamtest.run_pamtest(unix_username, "samba", [tc], [password, newpassword, newpassword], [password, newpassword, newpassword])
PamTestError: Error [2]: Test case { pam_operation [5] expected_rv [0] flags [0] } retured [28
]
PWRAP_DEBUG(29946) - pwrap_init: Initialize pam_wrapper
PWRAP_DEBUG(29946) - pwrap_init: Copy /usr/lib/x86_64-linux-gnu/libpam.so.0.83.1 to /tmp/pam.h/lib/libpam.so.0
PWRAP_DEBUG(29946) - copy_confdir: Copy config files from /home/mathieu/src/git.samba.org/samba/st/tmp/pam_services to /tmp/pam.h
PWRAP_DEBUG(29946) - pwrap_init: Successfully initialized pam_wrapper
command: /home/mathieu/src/git.samba.org/samba/python/samba/tests/test_pam_winbind_options.sh python /home/mathieu/src/git.samba.org/samba/bin/default/third_party/pam_wrapper/libpam-wrapper.so $SERVER $USERNAME $PASSWORD newp@ssword0 'try_authtok' 2>&1  | /home/mathieu/src/git.samba.org/samba/selftest/filter-subunit --fail-on-empty --prefix="samba.tests.pam_winbind with options 'try_authtok'(local)." --suffix="(ad_member)"
expanded command: /home/mathieu/src/git.samba.org/samba/python/samba/tests/test_pam_winbind_options.sh python /home/mathieu/src/git.samba.org/samba/bin/default/third_party/pam_wrapper/libpam-wrapper.so LOCALADMEMBER root loCalMemberPass newp@ssword0 'try_authtok' 2>&1  | /home/mathieu/src/git.samba.org/samba/selftest/filter-subunit --fail-on-empty --prefix="samba.tests.pam_winbind with options 'try_authtok'(local)." --suffix="(ad_member)"
ERROR: Testsuite[samba.tests.pam_winbind with options 'try_authtok'(local)(ad_member)]
REASON: Exit code was 1

 errors[1
```

**Comment by sathieu** [comment]
This PR includes a fix for travis. Please merge `autobuild: Fix travis build (broken since ldb lmdb backend)`: https://github.com/samba-team/samba/pull/168/commits/939c94832e722021222c4c145ce58406a0d23288

**Comment by cryptomilk** [comment]
`PWRAP_DEBUG(29945) - SYSLOG(7): username [LOCALADMEMBER/root] obtained`

This is obviously the wrong user :-)

You need to pass $DC_USERNAME and $DC_PASSWORD down in tests.py!

**Comment by cryptomilk** [comment]
Ah, this is the local test, however, root is the wrong user to use here!

Or are you running this as root? Which might be the problem :-)

**Comment by sathieu** [comment]
I'm running as root (inside a docker container). On travis, this is "LOCALADMEMBER/travis".

Should I create a user first? How?

**Comment by cryptomilk** [comment]
Have you tried only using domain credentials?

**Comment by sathieu** [comment]
Yes it fails too.

**Comment by sathieu** [comment]
I've improved the tests (creating a new user on the fly). But it's still unable to get echo_off strings:
```
PWRAP_DEBUG(31709) - SYSLOG(7): [pamh: 0x1156420] ENTER: pam_sm_chauthtok (flags: 0x4000)
PWRAP_DEBUG(31709) - SYSLOG(7): [pamh: 0x1156420] STATE: ITEM(PAM_SERVICE) = "samba" (0xfa0ac0)
PWRAP_DEBUG(31709) - SYSLOG(7): [pamh: 0x1156420] STATE: ITEM(PAM_USER) = "ADDOMAIN/TestPamOptionsUser" (0x10486f0)
PWRAP_DEBUG(31709) - SYSLOG(7): [pamh: 0x1156420] STATE: ITEM(PAM_CONV) = 0x1072ae0
PWRAP_DEBUG(31709) - SYSLOG(7): username [ADDOMAIN/TestPamOptionsUser] obtained
PWRAP_DEBUG(31709) - SYSLOG(7): getting password (0x00001021)
PWRAP_DEBUG(31709) - SYSLOG(5): could not recover authentication token
PWRAP_DEBUG(31709) - SYSLOG(7): unable to obtain a password
PWRAP_DEBUG(31709) - SYSLOG(5): password - (old) token not obtained
```

**Comment by sathieu** [comment]
I think I found an error: line 2227

    token = x_strdup(resp[i - replies].resp);

should be:

    token = x_strdup(resp[replies - 1].resp);

As the PAM_TEXT_INFO messages don't generate responses.

**Comment by cryptomilk** [comment]
Now you know why I want tests, to find and fix those bugs! :-)

**Comment by sathieu** [comment]
This is probably a pam_wrapper bug instead. As according to [pam_conv(3)](https://manpages.debian.org/stretch/libpam0g-dev/pam_conv.3.en.html):
> The index of the responses corresponds directly to the prompt index in the pam_message array.

**Comment by cryptomilk** [comment]
Could you try with: https://git.cryptomilk.org/users/asn/pam_wrapper.git/log/?h=master-pam_conv

**Comment by sathieu** [comment]
@cryptomilk It's not working. Currently testing this one (but I don't know how to test it in pam_wrapper):
```patch
diff --git a/src/libpamtest.c b/src/libpamtest.c
index 0a26c19..bd58fc7 100644
--- a/src/libpamtest.c
+++ b/src/libpamtest.c
@@ -214,7 +214,7 @@ static int pamtest_simple_conv(int num_msg,
 			       struct pam_response **response,
 			       void *appdata_ptr)
 {
-	int i, ri = 0;
+	int i = 0;
 	int ret;
 	struct pam_response *reply = NULL;
 	const char *prompt;
@@ -241,15 +241,12 @@ static int pamtest_simple_conv(int num_msg,
 
 			if (reply != NULL) {
 				if (prompt != NULL) {
-					ret = add_to_reply(&reply[ri], prompt);
+					ret = add_to_reply(&reply[i], prompt);
 					if (ret != PAM_SUCCESS) {
 						free_reply(reply, num_msg);
 						return ret;
 					}
-				} else {
-					reply[ri].resp = NULL;
 				}
-				ri++;
 			}
 
 			cctx->echo_off_idx++;
@@ -264,13 +261,12 @@ static int pamtest_simple_conv(int num_msg,
 
 			if (reply != NULL) {
 				if (prompt != NULL) {
-					ret = add_to_reply(&reply[ri], prompt);
+					ret = add_to_reply(&reply[i], prompt);
 					if (ret != PAM_SUCCESS) {
 						free_reply(reply, num_msg);
 						return ret;
 					}
 				}
-				ri++;
 			}
 
 			cctx->echo_on_idx++;
@@ -298,7 +294,7 @@ static int pamtest_simple_conv(int num_msg,
 		}
 	}
 
-	if (response && ri > 0) {
+	if (response) {
 		*response = reply;
 	} else {
 		free(reply);
```

**Comment by cryptomilk** [comment]
Could you try with https://git.cryptomilk.org/users/asn/pam_wrapper.git/log/?h=master-pam_conv again? I think I fixed it.

**Comment by cryptomilk** [comment]
Did you have time to try that branch?

**Comment by sathieu** [comment]
@cryptomilk Yes. It goes better and `pam_sm_chauthtok()` from `pam_winbind.c` returns success:
```
PWRAP_DEBUG(25673) - SYSLOG(7): [pamh: 0x555555e4a7f0] LEAVE: pam_sm_chauthtok returning 0 (PAM_SUCCESS)
PWRAP_DEBUG(25673) - SYSLOG(7): [pamh: 0x555555e4a7f0] STATE: ITEM(PAM_SERVICE) = "samba" (0x555555be98a0)
PWRAP_DEBUG(25673) - SYSLOG(7): [pamh: 0x555555e4a7f0] STATE: ITEM(PAM_USER) = "ADDOMAIN/TestPamOptionsUser" (0x555555e0bf10)
PWRAP_DEBUG(25673) - SYSLOG(7): [pamh: 0x555555e4a7f0] STATE: ITEM(PAM_OLDAUTHTOK) = 0x555555bea670
PWRAP_DEBUG(25673) - SYSLOG(7): [pamh: 0x555555e4a7f0] STATE: ITEM(PAM_CONV) = 0x555555be9970
PWRAP_DEBUG(25673) - SYSLOG(7): [pamh: 0x555555e4a7f0] STATE: DATA(PAM_WINBIND_LOGONSERVER) = "ADDC" (0x555555e549b0)
PWRAP_DEBUG(25673) - SYSLOG(7): [pamh: 0x555555e4a7f0] STATE: DATA(PAM_WINBIND_PWD_LAST_SET) = 0x5b3d2a9f
```

But this success is turned into a `PAM_MODULE_UNKNOWN` (28) by [pam_chauthtok() in pam_password.c](https://sources.debian.org/src/pam/1.1.8-3.7/libpam/pam_password.c/?hl=12#L12) probably from [_pam_dispatch_aux() in pam_dispacth.c](https://sources.debian.org/src/pam/1.1.8-3.7/libpam/pam_dispatch.c/?hl=104#L104)

File: python/samba/tests/pam_winbind_chauthtok.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+class PamChauthtokTests(samba.tests.TestCase):
+    def test_chauthtok(self):
+        domain = os.environ["DOMAIN"]
+        username = os.environ["USERNAME"]
+        password = os.environ["PASSWORD"]
+        newpassword = os.environ["NEWPASSWORD"]
+        pam_options = os.environ["PAM_OPTIONS"]
+        unix_username = "%s/%s" % (domain, username)
+        expected_rc = 0 # PAM_SUCCESS
+
```
Comment (cryptomilk) [inline python/samba/tests/pam_winbind_chauthtok.py]: Don't we have to set the pam data for PAM_AUTHTOK here?

`os.environ["PAM_AUTHTOK"] = password`

Then the pam_set_items.so module will hand it to pam_winbind.

**Comment by cryptomilk** [comment]
I've got the test working:

https://git.samba.org/?p=asn/samba.git;a=shortlog;h=refs/heads/master-pwrap

**Comment by cryptomilk** [comment]
I will do a new pam_wrapper release and then push it upstream.

**Comment by sathieu** [comment]
Thanks!

Please go ahead!

**Comment by cryptomilk** [comment]
See https://gitlab.com/samba-team/samba/merge_requests/82

**Comment by sathieu** [comment]
Fixed by 62400224. Great!

**Pull request closed without merging**